quinn-proto/src/assembler.rs

```diff
@@ -71,6 +74,32 @@ impl Assembler {
         }
     }
 
+    // Copy the buffered chunk data to new chunks backed by a single buffer to
+    // make sure we're not unnecessarily holding on to many larger allocations.
+    // Merge contiguous chunks in the process of doing so. Reset the `defragmented`
+    // counter to the new number of chunks left in the heap so that we can decide
+    // when to defragment the queue again if necessary.
```

Comment:
Should this skip previously-defragmented chunks? Little or no space is saved by reprocessing them, and it might substantially reduce the amount of work required by a highly fragmented stream with a large buffer and a slow application, since the unread prefix would otherwise be copied on each defragment pass. Should be easy to do by e.g. `take`ing the first `self.defragmented` elements from `old.into_sorted_vec().into_iter()` before proceeding.

Reply:
I thought about that a little bit, but the problem is you're not at all sure that the already-defragmented chunks are a prefix of the heap. At least, I think they can come in in basically arbitrary order as long as they stay within flow credit, right?

Reply:
Ah, good point. We could accomplish it with a flag bit, or by storing defragmented chunks separately, but the complexity tradeoff vs. a constant factor improvement for a weird case is less clear for those.

Reply:
I've been thinking a bit more about this. I think the current behavior is even desirable: if the malicious peer continues his sending of 1-byte frames, then eventually we are better off if we redefragment all the chunks rather than only defragmenting only once. Also, the extra state would make the performance in the fast/usual case somewhat worse again; so optimizing the bad/uncommon case that way doesn't seem like the right trade-off.